Improve Makefile (install, uninstall)

xexyl opened the pull request:
Add to the Makefile the PREFIX variable so one can change the install location from the default /usr/local. For instance to use /usr:

    make PREFIX=/usr install

Added uninstall rule to uninstall the programs, library and man pages. If one specified a different PREFIX in install they would have to specify it for uninstall too like:

    make PREFIX=/usr uninstall

To be congruent with install, the uninstall allows one to silence the rm verbosity (RM_V variable instead of INSTALL_V variable).

Updated README.md file for the above changes (well not the RM_V part).

xexyl committed:
Improve Makefile (install, uninstall)

Add to the Makefile the PREFIX variable so one can change the install
location from the default /usr/local. For instance to use /usr:

    make PREFIX=/usr install

Added uninstall rule to uninstall the programs, library and man pages.
If one specified a different PREFIX in install they would have to
specify it for uninstall too like:

    make PREFIX=/usr uninstall

To be congruent with install, the uninstall allows one to silence the rm
verbosity (RM_V variable instead of INSTALL_V variable).

Updated README.md file for the above changes (well not the RM_V part).

xexyl commented:
Perhaps you'd like to sync this repo, the dyn_array repo and the jparse repo to the mkiocccentry subdirectories? I must leave now .. have things I have to do and they'll take a while: might not even finish it today.

lcn2 commented:
# QUESTION 

Should instances of `/usr/local` in the `make depend` rule become `${PREFIX}` ?

lcn2 reviewed (changes requested):
See [GH-pullrequestreview-2274502182](https://github.com/lcn2/dbg/pull/22#pullrequestreview-2274502182).

lcn2 reviewed (changes requested):
# QUESTION 

Should instances of `/usr/local` in the `make depend` rule become `${PREFIX}` ?    

## UPDATE 0

You might wish to modify the `make depend` llne from:

```make
        ${SED} -E -e 's;\s/usr/local/include/\S+;;g' -e 's;\s/usr/include/\S+;;g' | \
```

to:

```make
        ${SED} -E -e 's;\s${PREFIX}/include/\S+;;g' -e 's;\s/usr/include/\S+;;g' | \
```

BTW: It is OK if "_PREFIX == /usr_" as the 2nd sed expression would just become redundant.  And if "_PREFIX != /usr_" then the 2nd `/usr/include` related sed expression is still needed to address default system includes.

## UPDATE 1

Please also consider [GH-pullrequestreview-2274499706](https://github.com/ioccc-src/mkiocccentry/pull/953#pullrequestreview-2274499706).  We will hold off in processing this PR so that you may more easily update and test the combined set of 3 PRs across the 3 impacted repos.

## UPDATE 2

Overall we like the concept you have introduced, @xexyl.  With the above suggested modifications as well as solving the check failure issues, the modified PR will be a welcome addition.

xexyl commented:
Good point. I don't think it's necessary as I don't think macOS adds that with -MM but I could be wrong. 

I will have to do this tomorrow. 

Sorry for the delays there but it will be worth it I think.

xexyl committed:
Add LD_SPECIAL variable to Makefile

The default value is empty but LDFLAGS is set to ${LD_SPECIAL} for
special cases.

xexyl commented:
I don't think the make depend should be or needs to be modified. The `${CC}` doesn't need or use the `${PREFIX}` and macOS does not add with `-MM` the directory `/usr/include` either. Given the problems in the mkiocccentry Makefile that I mentioned (the `LD_SPECIAL` does not work as you hoped and the `C_SPECIAL` variable is also not what you hoped if more than one option has to be specified) perhaps this could be merged and you can then play with this more? This way at least the `PREFIX` functionality is there. But since you want to process all three at the same time and since that one won't work as you hoped ...

Updating the depend rule to include `PREFIX` does nothing because:

```sh
mkdir foo ; cp dbg.h foo/
make PREFIX=foo depend
```

what then? The `Makefile` won't change as it stands. This is because the compiler does not have the include directories updated since the depend rule uses:

```makefile
${CC} ${CFLAGS} -MM -I. ${ALL_CSRC}
```

and the `${CFLAGS}` does not include `PREFIX`. 

Please advise how you wish me to proceed or if you want you can merge this and then play with it.

lcn2 commented:
We will process these PRs as is and then we will fix, if and as needed afterwards.
